packages/models-library/tests/test_utils_string_substitution.py

```diff
@@ -162,15 +164,87 @@ def test_substitution_with_new_and_legacy_identifiers():
     ids=lambda p: f"{p.parent.name}/{p.name}",
 )
 def test_substitution_against_service_metadata_configs(metadata_path: Path):
-
     meta_str = metadata_path.read_text()
     meta_str = substitute_all_legacy_identifiers(meta_str)
 
-    template = TemplateText(meta_str)
+    template = TextTemplate(meta_str)
     assert template.is_valid()
 
     found = template.get_identifiers()
     if found:
         assert all(
             identifier in KNOWN_IDENTIFIERS for identifier in found
         ), f"some identifiers in {found} are new and therefore potentially unsupported"
+
+
+def test_template_substitution_on_envfiles():
+    envfile_template = dedent(
+        """
+    x=$VALUE1
+    y=$VALUE2
+    """
+    )
+    template = TextTemplate(envfile_template)
+    assert set(template.get_identifiers()) == {"VALUE1", "VALUE2"}
+
+    # NOTE how it casts string to to int
+    assert template.substitute({"VALUE1": "3", "VALUE2": 3}) == dedent(
+        """
+    x=3
+    y=3
+    """
+    )
+
+    # NOTE does not cast if it is in a container
+    assert template.substitute({"VALUE1": ["3", "4"], "VALUE2": [3, 4]}) == dedent(
+        """
+    x=['3', '4']
+    y=[3, 4]
+    """
```

Comment:
From what I get here, python values are serialised to string. Did I get it right?

Reply:
its not strictly about serialization but about substitution